mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -772,29 +798,100 @@ static void reconfigureLAGeneric(LinalgAlignRewriter &b,
+/// that if the inputs all fall into the padding, the result of the elementwise
+/// function will also be the expected zero.
+static LogicalResult knownToPreserveZero(linalg::GenericOp laGeneric,
+                                         LinalgAlignRewriter &b) {
+  // Brute-force test: clone the generic, replace all the arguments with 0s,
+  // and constant-fold.
+  LLVM_DEBUG(llvm::dbgs() << "* Cloning generic to test if it allows 0s\n");
+  auto clonedOp = cast<linalg::GenericOp>(b.clone(*laGeneric));
+  Location loc = clonedOp.getLoc();
+  LinalgAlignRewriter::InsertionGuard guard(b);
+  b.setInsertionPointToStart(&clonedOp.getRegion().front());
+  OperationFolder folder(clonedOp.getContext(), b.getListener());
+  for (BlockArgument &arg : clonedOp.getRegion().getArguments()) {
+    Value zero = createZeroConstantOp(b, loc, arg.getType());
+    arg.replaceAllUsesWith(zero);
+  }
+  for (Operation &op :
+       llvm::make_early_inc_range(clonedOp.getRegion().getOps())) {
+    bool ignored = false;
+    (void)folder.tryToFold(&op, &ignored);
   }
-  return success(isLegal);
+  LLVM_DEBUG(llvm::dbgs() << "* Folded input fusion region to " << clonedOp
+                          << "\n");
+  auto yieldOp =
+      cast<linalg::YieldOp>(clonedOp.getRegion().front().getTerminator());
+  bool foldedToZero = llvm::all_of(yieldOp.getValues(), [&](Value v) {
+    return matchPattern(v, m_AnyZeroFloat()) || matchPattern(v, m_Zero());
+  });
+  LLVM_DEBUG(llvm::dbgs() << "* Cloning generic to test if it allows 0s\n");
+  b.eraseOp(clonedOp);
+  return success(foldedToZero);
+}
+
+/// If this generic doesn't preserve zero (ex, it's x => x + 1) and if the
+/// validity of the tiling operation was being tracked (this indicates input
+/// fusion), then:
+/// - Clone the output tile
+/// - Set up a register->register threadwise_read_into between this cloned tile
+///   and the original output tile, with dynamic validities drawn from the
+///   validity results of each read.
+/// The extra threadwise_read_into we create here will cause elements that
+/// didn't actually get fetched from memory to become 0s again thanks to an if
+/// statement in what would otherwise be a memcpy().
+///
+/// Returns the validity record from the padding read if there is one.
+static Value reapplyPaddingIfNeeded(linalg::GenericOp reconfiguredGeneric,
+                                    ValueRange validityRecords,
+                                    ThreadwiseReadIntoOp oldTwRead,
+                                    LinalgAlignRewriter &b) {
+  // If the old read isn't recording its validity anywhere, or if we're not
+  // using that record, we don't need to propagate the validity records from the
+  // inputs.
+  if (!oldTwRead.getValidityRecord() ||
+      oldTwRead.getValidityRecord().use_empty()) {
+    if (validityRecords.empty())
+      return Value{};
+    if (succeeded(knownToPreserveZero(reconfiguredGeneric, b)))
+      return Value{};
```

Comment:
maybe lets do FailureOr<Value> as the return type.
It d play nicely with emitting errors below...